src/DependencyInjection/Compiler/FragmentRegistryPass.php

```diff
@@ -0,0 +1,106 @@
+     * tagged services to the fragment registry.
+     *
+     * @param ContainerBuilder $container
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->has('contao.fragment_registry')) {
+            return;
+        }
+
+        $fragmentRegistry = $container->findDefinition('contao.fragment_registry');
+
+        if (!$this->classImplementsInterface($fragmentRegistry->getClass(), FragmentRegistryInterface::class)) {
+            return;
+        }
+
+        // Search type providers
+        $typeProviders = $container->findTaggedServiceIds('contao.fragment_types_provider');
+
+        foreach ($typeProviders as $id => $tags) {
+
+            $fragmentTypeProvider = $container->findDefinition($id);
+
+            if (!$this->classImplementsInterface($fragmentTypeProvider->getClass(), FragmentTypesProviderInterface::class)) {
+                throw new LogicException(sprintf('The class "%s" was registered as "contao.fragment_types_provider" but does not implement the interface "%s".',
+                    $fragmentTypeProvider->getClass(),
+                    FragmentTypesProviderInterface::class
+                ));
+            }
+
+            // Resolve the fragment type provider to ask for the types
+            /* @var $fragmentTypeProviderInstance $fragmentTypeProvider */
+            $fragmentTypeProviderInstance = $container->resolveServices($fragmentTypeProvider);
+
+            foreach ($fragmentTypeProviderInstance->getFragmentTypes() as $fragmentTypeInterface => $tag) {
+
+                // Register the type
+                $fragmentRegistry->addMethodCall('addFragmentType', [$fragmentTypeInterface]);
+
+                $taggedServices = $container->findTaggedServiceIds($tag);
+
+                foreach ($taggedServices as $id => $tags) {
+
+                    $fragment = $container->findDefinition($id);
+
+                    if (!$this->classImplementsInterface($fragment->getClass(), $fragmentTypeInterface)) {
+                        throw new LogicException(sprintf('The class "%s" was registered as "%s" but does not implement the interface "%s".',
+                            $fragment->getClass(),
+                            $tag,
+                            $fragmentTypeInterface
+                        ));
+                    }
+
+                    $fragmentRegistry->addMethodCall('addFragment', [new Reference($id)]);
+                }
+            }
+
+            // Set provider to private so it's removed during optimization routine
+            $fragmentTypeProvider->setPublic(false);
+            unset($fragmentTypeProviderInstance);
+        }
+    }
+
+    /**
+     * Checks if a given class name implements a given interface name.
+     *
+     * @param string $class
+     * @param string $interface
+     *
+     * @return bool
+     */
+    private function classImplementsInterface($class, $interface)
```

Comment:
`is_a()` did not check for inherited interfaces for me. Reflection is better here.